yapapi/script/__init__.py

```diff
@@ -0,0 +1,172 @@
+import asyncio
+from datetime import timedelta
+from typing import Any, Awaitable, Callable, Dict, Optional, List, Tuple, TYPE_CHECKING
+
+from yapapi.events import CommandExecuted
+from yapapi.script.capture import CaptureContext
+from yapapi.script.command import (
+    BatchCommand,
+    Command,
+    Deploy,
+    DownloadBytes,
+    DownloadFile,
+    DownloadJson,
+    Run,
+    SendBytes,
+    SendFile,
+    SendJson,
+    Start,
+    Terminate,
+)
+from yapapi.storage import DOWNLOAD_BYTES_LIMIT_DEFAULT
+
+if TYPE_CHECKING:
+    from yapapi.ctx import WorkContext
+
+
+class Script:
+    """Represents a series of commands to be executed on a provider node.
+
+    New commands are added to the script either through its `add` method or by calling one of the
+    convenience methods provided (for example: `run` or `send_json`).
+    Adding a new command *does not* result in it being immediately executed. Once ready, a `Script`
+    instance is meant to be yielded from a worker function (work generator pattern).
+    Commands will be run in the order in which they were added to the script.
+
+    If the `WorkContext` instance this `Script` uses has the field `_implicit_init` set to `True`,
+    the first script to be yielded is going to prepend the user's commands with `Deploy` and
+    `Start` commands.
+    """
+
+    timeout: Optional[timedelta] = None
+    """Time after which this script's execution should be forcefully interrupted."""
+
+    wait_for_results: bool = True
+    """Whether this script's execution should block until its results are available."""
+
+    def __init__(self, context: "WorkContext"):
+        self._ctx: "WorkContext" = context
+        self._commands: List[Tuple[Command, asyncio.Future]] = []
+
+    def _evaluate(self) -> List[BatchCommand]:
+        """Evaluate and serialize this script to a list of batch commands."""
+        batch: List[BatchCommand] = []
+        for cmd, _future in self._commands:
+            batch.append(cmd.evaluate(self._ctx))
+        return batch
+
+    async def _after(self):
+        """Hook which is executed after the script has been run on the provider."""
+        for cmd, _future in self._commands:
+            await cmd.after(self._ctx)
```

Comment:
Wouldn't 
```
await asyncio.gather(cmd.after(self._ctx) for cmd, _ in self._commands)
```
be better?

Reply:
Define "better". :D This is subjective, but I find the current version easier to read. Does using `gather` give us a benefit over a for-loop?

Reply:
Well, it obviously does give us a performance benefit. E.g. if we have two `after`s, and first one must wait for some external resource, with `gather` the second will finish while the first is waiting.

So I'm sure `gather` is better **unless** we care about the order of execution. I think we shouldn't care, but I'm not sure about this.

Reply:
Let's say we are running two commands on the provider. Should we run `after` in the same order that those commands were run? If yes, then let's not switch to `asyncio.gather`.

Reply:
Since both `before` and `after` hooks on `Command`s may interact directly with the remote node (i.e. the provider) I'd be against making them execute in parallel. This may lead to unwanted/unexpected behaviour, especially since the script commands themselves are intended to be run sequentially.

Reply:
Do we *really* want `before`/`after` to interact with the remote node?
If yes, what is the purpose of the `before`/`after` separation?

I won't insist, but I think that if we can restrict this, we should.

Reply:
if the execution of the before / after hooks for subsequent commands should depend on the completion of those hooks for previous commands (which I'm just plain unsure of), then I guess using `gather` is not possible... 

anyway, since the execution had been sequential in the previous implementation and no-one complained about performance of this particular feature, I'd leave the for-loop here